README.md
# viu-player


## [online demo](http://blog.gemer.xyz/viu-player/)  
## Table of Contents

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Installation](#installation)
- [Usage](#usage)
  - [`<script>` Tag](#script-tag)
  - [Browserify/CommonJS](#browserifycommonjs)
  - [RequireJS/AMD](#requirejsamd)
  - [字幕面板](#%E5%AD%97%E5%B9%95%E9%9D%A2%E6%9D%BF)
  - [下一集播放器事件](#%E4%B8%8B%E4%B8%80%E9%9B%86%E6%92%AD%E6%94%BE%E5%99%A8%E4%BA%8B%E4%BB%B6)
  - [双字幕开关通事件](#%E5%8F%8C%E5%AD%97%E5%B9%95%E5%BC%80%E5%85%B3%E9%80%9A%E4%BA%8B%E4%BB%B6)
  - [获取进条度是否可以快进快退](#%E8%8E%B7%E5%8F%96%E8%BF%9B%E6%9D%A1%E5%BA%A6%E6%98%AF%E5%90%A6%E5%8F%AF%E4%BB%A5%E5%BF%AB%E8%BF%9B%E5%BF%AB%E9%80%80)
  - [主动触发信息按键](#%E4%B8%BB%E5%8A%A8%E8%A7%A6%E5%8F%91%E4%BF%A1%E6%81%AF%E6%8C%89%E9%94%AE)
  - [语言选择格式](#%E8%AF%AD%E8%A8%80%E9%80%89%E6%8B%A9%E6%A0%BC%E5%BC%8F)
  - [css格式状态](#css%E6%A0%BC%E5%BC%8F%E7%8A%B6%E6%80%81)
  - [加载超时机制](#%E5%8A%A0%E8%BD%BD%E8%B6%85%E6%97%B6%E6%9C%BA%E5%88%B6)
  - [info 传入字段](#info-%E4%BC%A0%E5%85%A5%E5%AD%97%E6%AE%B5)
  - [advance 高级配置 传入字段](#advance-%E9%AB%98%E7%BA%A7%E9%85%8D%E7%BD%AE-%E4%BC%A0%E5%85%A5%E5%AD%97%E6%AE%B5)
  - [](#)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Installation

```sh
npm install --save viu-player
```

## Usage

To include viu-player on your website or web application, use any of the following methods.

### `<script>` Tag

This is the simplest case. Get the script in whatever way you prefer and include the plugin _after_ you include [video.js][videojs], so that the `videojs` global is available.

```html
<script src="//path/to/video.min.js"></script>
<script src="//path/to/viu-player.min.js"></script>
<script>
  var player = videojs("my-video");

  player.ViuPlayer();
</script>
```

### Browserify/CommonJS

When using with Browserify, install viu-player via npm and `require` the plugin as you would any other module.

```js
var videojs = require("video.js");

// The actual plugin function is exported by this module, but it is also
// attached to the `Player.prototype`; so, there is no need to assign it
// to a variable.
require("viu-player");

var player = videojs("my-video");

player.ViuPlayer();
```

### RequireJS/AMD

When using with RequireJS (or another AMD library), get the script in whatever way you prefer and `require` the plugin as you normally would:

```js
require(["video.js", "viu-player"], function(videojs) {
  var player = videojs("my-video");

  player.ViuPlayer();
});
```

### 字幕面板

```js

player.on("subtitleChange", function(event,langId) {
  //data :string etc: 1,2,3 langId
});
player.on("subtitleClick", function(event,langData) {
  //data :string langData
});

player.ViuPlayer().loadNewLangListOptions(langData);
```

### 下一集播放器事件

```js
 //主动方法
player.ViuPlayer()toggleNextChip();

player.on("nextChip", function() {
  //
});
```

### 双字幕开关通事件

```js
   //隐藏方法
   player.ViuPlayer().hideTopSubtitle();   
   // 打开方法
   player.ViuPlayer().showTopSubtitle();
   player.on("toggleTopSubtitle", function(event, switcher) {
      console.log("收到双字幕开关", switcher);
  })
```

### 获取进条度是否可以快进快退
```js
   player.ViuPlayer().checkVideoProcessed();

```

### 主动触发信息按键
```js
   player.ViuPlayer().toggleInfo();

```

### 语言选择格式

```json
languages: [
    {
      id: 'CN',
      name: '简体中文',
      isDefault: 0
    },
    {
      id: 'EN',
      name: 'ENGLISH',
      isDefault: 1
    },
    {
      id: 'JP',
      name: '日文',
      isDefault: 0
    }
  ],
```

### css格式状态
 ```css
 .lang-active{
   // 语言选择状态
 }
 .status-active{
  // 六图标状态
 }

 ```

 ### 加载超时机制
```js
   examplePlayer.on("loadTimeout", function() {
      console.log("loading超60秒");
    });

  options.videoBufferTimeout  // 定义网络慢超时时间

```


 ### info 传入字段
```js
    info: {
    chipName: '',
    episode: '',
    chipCaptions: '',
    chipDesc: '',
    chipTotal: 0,
    isMovie: 0,
    chipActiveTotal: 0,
    userLevel: '',
    pUrl: '',
    showPremium: true,
  }
```

 ### advance 高级配置 传入字段
```js
  advance: {
    resolutionList: ['s1080p', 's720p', 's480p', 's240p'],
    defaultResolution: 's720p',
    cdnList: ['url', 'url2', 'url3'],
    defaultCdn: 'url2'
  }
```
 ### 
```js
   examplePlayer.on("viuAdvance", function(event,data) {
      console.log("viuAdvance",data);  
      // ex: selectId: s1080p
    });

   examplePlayer.on("weakNetwork", function(event,data) {
     // 网络提示
      //
    });
```
## License

MIT. Copyright (c) gemer cheung &lt;gemercheung@gmail.com&gt;

[videojs]: http://videojs.com/


CHANGELOG.md
<a name="0.4.2"></a>
## [0.4.2](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.4.2-nc.14...v0.4.2) (2020-08-24)

<a name="0.4.1"></a>
## [0.4.1](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.4.1-nc.3...v0.4.1) (2020-08-10)

<a name="0.4.0"></a>
# [0.4.0](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.11...v0.4.0) (2020-07-27)

<a name="0.3.11"></a>
## [0.3.11](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.10...v0.3.11) (2020-06-04)

<a name="0.3.10"></a>
## [0.3.10](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.9...v0.3.10) (2020-06-02)

<a name="0.3.9"></a>
## [0.3.9](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.8...v0.3.9) (2020-06-01)

<a name="0.3.8"></a>
## [0.3.8](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.7...v0.3.8) (2020-06-01)

<a name="0.3.7"></a>
## [0.3.7](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.6...v0.3.7) (2020-05-12)

<a name="0.3.6"></a>
## [0.3.6](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.5...v0.3.6) (2020-04-27)

<a name="0.3.5"></a>
## [0.3.5](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.4...v0.3.5) (2020-04-21)

<a name="0.3.4"></a>
## [0.3.4](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.3...v0.3.4) (2020-04-20)

<a name="0.3.3"></a>
## [0.3.3](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.2...v0.3.3) (2020-04-17)

<a name="0.3.2"></a>
## [0.3.2](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.1...v0.3.2) (2020-04-17)

<a name="0.3.1"></a>
## [0.3.1](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.3.0...v0.3.1) (2020-04-16)

<a name="0.3.0"></a>
# [0.3.0](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.12...v0.3.0) (2020-03-02)

<a name="0.2.12"></a>
## [0.2.12](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.11...v0.2.12) (2020-02-27)

<a name="0.2.11"></a>
## [0.2.11](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.10...v0.2.11) (2020-01-15)

<a name="0.2.10"></a>
## [0.2.10](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.7...v0.2.10) (2019-12-12)

<a name="0.2.9"></a>
## [0.2.9](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.8...v0.2.9) (2019-12-11)

<a name="0.2.7"></a>
## [0.2.7](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.4...v0.2.7) (2019-12-10)

<a name="0.2.6"></a>
## [0.2.6](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.5...v0.2.6) (2019-12-10)

<a name="0.2.5"></a>
## [0.2.5](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.4...v0.2.5) (2019-12-10)

<a name="0.2.4"></a>
## [0.2.4](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.2...v0.2.4) (2019-12-10)

<a name="0.2.3"></a>
## [0.2.3](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.2...v0.2.3) (2019-12-04)

<a name="0.2.2"></a>
## [0.2.2](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.1...v0.2.2) (2019-11-29)

<a name="0.2.1"></a>
## [0.2.1](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.2.0...v0.2.1) (2019-11-28)

<a name="0.2.0"></a>
# [0.2.0](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.57...v0.2.0) (2019-11-28)

<a name="0.1.58"></a>
## [0.1.58](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.57...v0.1.58) (2019-11-28)

<a name="0.1.57"></a>
## [0.1.57](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.56...v0.1.57) (2019-11-27)

<a name="0.1.56"></a>
## [0.1.56](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.55...v0.1.56) (2019-11-27)

<a name="0.1.55"></a>
## [0.1.55](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.53...v0.1.55) (2019-11-27)

<a name="0.1.54"></a>
## [0.1.54](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.53...v0.1.54) (2019-11-26)

<a name="0.1.53"></a>
## [0.1.53](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.52...v0.1.53) (2019-11-26)

<a name="0.1.52"></a>
## [0.1.52](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.45...v0.1.52) (2019-11-26)

<a name="0.1.51"></a>
## [0.1.51](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.50...v0.1.51) (2019-11-22)

<a name="0.1.50"></a>
## [0.1.50](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.49...v0.1.50) (2019-11-22)

<a name="0.1.49"></a>
## [0.1.49](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.48...v0.1.49) (2019-11-20)

<a name="0.1.48"></a>
## [0.1.48](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.47...v0.1.48) (2019-11-19)

<a name="0.1.47"></a>
## [0.1.47](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.46...v0.1.47) (2019-11-19)

<a name="0.1.46"></a>
## [0.1.46](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.41...v0.1.46) (2019-11-18)

<a name="0.1.45"></a>
## [0.1.45](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.44...v0.1.45) (2019-11-18)

<a name="0.1.44"></a>
## [0.1.44](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.43...v0.1.44) (2019-11-17)

<a name="0.1.43"></a>
## [0.1.43](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.42...v0.1.43) (2019-11-14)

<a name="0.1.42"></a>
## [0.1.42](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.40...v0.1.42) (2019-11-14)

<a name="0.1.41"></a>
## [0.1.41](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.40...v0.1.41) (2019-11-13)

<a name="0.1.40"></a>
## [0.1.40](https://bitbucket.org/viugzteamwebos/videojs-viu-tools/compare/v0.1.39...v0.1.40) (2019-11-12)

<a name="0.1.39"></a>

demo/demo.js
(function (window, videojs) {
  const examplePlayer = (window.examplePlayer = videojs(
    "videojs-viu-tools-player",
    {
      preload: "auto",
      autoplay: true,
      fluid: false,
      inactivityTimeout: 0, // 非用户活跃时间间隔,在keydown事件控制非用户活跃
      errorDisplay: true,
      controlBar: false,
      loop: true,
      muted: false,
      html5: {
        hls: {
          overrideNative: true,
          enableLowInitialPlaylist: true,
          smoothQualityChange: true,
          useBandwidthFromLocalStorage: true,
        },
        nativeTextTracks: false,
        nativeAudioTracks: false,
        nativeVideoTracks: false,
      },
      language: "zh_cn",
      languages: {
        zh_cn: {
          hide_sub: "隐藏",
          episode_num: "第13集",
          setting_language: "语言 / Language",
          off_shelf_date:"下架日期",
          resolution_title:"源大小",
          fontsize_s: "小",
          fontsize_m: "正常",
          fontsize_l: "大",
          fontsize_xl:"超大",
          resolution_hero_title: "高级设置",
          cdn_title:"CDN线路",
          res_s1080p:"超清10M",
          res_s720p:"超清4M",
          res_s480p:"高清",
          res_s240p:"清晰",
          res_url:"Amazon",
          res_url2:"PCCW HKT",
          res_url3:"AKimai",

        },
        en: {
          hide_sub: "Close",
          episode_num: "第13集",
          setting_language: "Language / 语言",
          video_timeout: "Please connect to a better network",
          fontsize_s: "small",
          fontsize_m: "normal",
          fontsize_l: "large",
          resolution_title:"源大小",
         
        },
        sg_en: {
          hide_sub: "Close",
          episode_num: "EP. 13",
          setting_language: "Language / 语言",
          off_shelf_date:"下架日期 : "
        },
        zh_hk: {
          hide_sub: "隱藏",
          episode_num: "第13集",
          setting_language: "語言 / Language",
        },
        au_en: {
          hide_sub: "Close",
          episode_num: "EP. 13",
          setting_language: "Language / 语言",
        },
        th: {
          hide_sub: "ปิด",
          episode_num: "EP. 13",
          setting_language: "การตั้งค่าภาษา",
        },
        ph_en: {
          hide_sub: "Close",
          episode_num: "EP. 13",
          setting_language: "Language / 语言",
        },
        nz_en: {
          hide_sub: "Close",
          episode_num: "EP. 13",
          setting_language: "Language / 语言",
        },
      },
    }
  ));
  const ViuPlayer = (window.ViuPlayer = examplePlayer.ViuPlayer({
    isTopSubtitleDisable: true,
    fontSize: ["s", "m", "l", "xl"],
    defaultFontSize: "s",
    info: {
      chipName: "生活大爆作 第十三季",
      episode: "12",
      chipCaptions: "谢耳朵与艾米终于有自己的孩子呢...",
      chipDesc: "谢耳朵与艾米 经历了很多事，终于................ ",
      offlineTime: "",
      offlineTimeFormat: "YYYY-MM-DD",
      isMovie: 0,
    },
    advance: {
      resolutionList: ["s1080p", "s720p", "s480p", "s240p"],
      defaultResolution: 's480p',
      cdnList: ['url','url2','url3'],
      defaultCdn: 'url',
      isCDNOpen: true,
    },
    playToggle: {
      replay: false,
    },
    controlBar: {
      children: [
        "currentTimeDisplay",
        "progressControl",
        "durationDisplay",
        "liveDisplay",
        "seekToLive",
      ],
    },
    videoBufferTimeout: 8,
  }));

  examplePlayer.src(
    "https://bitdash-a.akamaihd.net/content/MI201109210084_1/m3u8s/f08e80da-bf1d-4e3d-8899-f0f6155f6efa.m3u8"
  );

  examplePlayer.on("nextChip", function (e) {
    console.log("收到下一集");
  });
  examplePlayer.on("toggleTopSubtitle", function (event, switcher) {
    console.log("收到双字幕开关", switcher);
  });

  examplePlayer.on("subtitleChange", function (event, data) {
    console.log("subtitleChange data", data);
  });

  examplePlayer.on("loadTimeout", function () {
    console.log("loading超60秒");
  });
  examplePlayer.on("viuAdvance", function (event, data) {
    console.log("viuAdvance-selected", data);
  });


  window.addEventListener("keydown", function (event) {
    console.log("keyCode", event.keyCode);
    switch (event.keyCode) {
      case 13:
        examplePlayer.ViuPlayer().confirm();
        break;
      case 37:
        examplePlayer.ViuPlayer().leftSelectAction();
        break;
      case 38:
        examplePlayer.ViuPlayer().upperSelectAction();
        break;
      case 39:
        examplePlayer.ViuPlayer().rightSelectAction();
        break;
      case 40:
        examplePlayer.ViuPlayer().downerSelectAction();
        break;
      case 66: {
        // b back
        examplePlayer.ViuPlayer().return();
        break;
      }
      case 73: {
        // s 字幕 for test
        examplePlayer.ViuPlayer().toggleInfo();
        break;
      }
      case 83: {
        // s 字幕 for test

        examplePlayer.ViuPlayer().toggleSubtitle();
        break;
      }
      case 80: {
        // p 禁用双字幕

        examplePlayer.ViuPlayer().checkTopSubtitleDisable(false);
        break;
      }
      case 79: {
        // o 开启双字幕

        examplePlayer.ViuPlayer().checkTopSubtitleDisable(true);
        break;
      }

      default:
        break;
    }
  });

  // setTimeout(function () {
  //   ViuPlayer.toggleNextChip()
  // }, 3000)
})(window, window.videojs);
